### PR TITLE
Allow for excluding data from the metrics files

### PR DESCRIPTION
**Pull request by jarretlavallee**
Prior to this PR, there was no mechanism for filtering out metrics
from the json files. This PR adds filter capabilities and defaults
to removing the `pe-puppet-profiler` and `file-sync-storage-service`
hashes from the `puppetserver` metrics.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -20,6 +20,9 @@
   Optional[Integer] $metrics_server_port      = undef,
   Optional[String]  $metrics_server_db_name   = undef,
   Optional[String]  $override_metrics_command = undef,
+  Optional[Array[String]] $puppetserver_excludes = ['pe-puppet-profiler', 'file-sync-storage-service'],
```
Comment (Sharpie) [inline manifests/init.pp]: We could also exclude the `pe-master`, and `pe-jruby-metrics` hashes as well. They're not nearly as big as `pe-puppet-profiler`, but worth excluding as well.

Also, this should default to just File Sync, `['file-sync-storage-service']`, if the PE version is older than 2017.3 --- as the `pe-` metrics were the only source of the profiling data prior to Puppet Server 5.0.

File: files/tk_metrics [hunk 2]
```diff
@@ -171,7 +185,8 @@ HOSTS.each do |host|
     dataset['servers'][hostkey][METRICS_TYPE]['api-query-start'] = timestamp.utc.iso8601
     dataset['servers'][hostkey][METRICS_TYPE]['api-query-duration'] = Time.now - timestamp
 
-    json_dataset = JSON.pretty_generate(dataset)
+    filtered_dataset = filter_metrics(dataset, EXCLUDES)
+    json_dataset = JSON.pretty_generate(filtered_dataset)
```
Comment (Sharpie) [inline files/tk_metrics]: I did some benchmarking and it turns out that dropping the extra whitespace created by `pretty_generate` is going to save us significant amounts of space.

For a monolithic master with 8 compilers and one deployed Code Manager environment, the daily archive size was 16 MB. Removing the `file-sync-storage-service` and `pe-` datasets drops that to 13 MB.

Removing the whitespace created by `pretty_generate` and keeping the data drops the archive size to 7.8 MB.

Removing both the whitespace and the data drops it to 5.6 MB. Disk space required for uncompressed files was reduced from 200 MB to 50 MB once everything was removed.

I would expect the savings from removing `file-sync-storage-service` to increase rapidly as the number of environments grows, but killing off the whitespace is still a significant reduction in size.

File: manifests/init.pp (hunk 1 above)
Comment (jarretlavallee) [reply manifests/init.pp]: Good call. I'll put in some logic to determine the defaults and allow for future version changes.

File: files/tk_metrics (hunk 2 above)
Comment (jarretlavallee) [reply files/tk_metrics]: My only concern here is the ability to grep metrics without using a post-processing tool. The data is the same, but the README instructions will need to change, so I guess this could be considered a breaking change? 

I can update the instructions to use something like `jq` and the minified JSON though.  Would it be better to add instructions with `jq` or do something with ruby/another built-in library?

File: manifests/init.pp (hunk 1 above)
Comment (jarretlavallee) [reply manifests/init.pp]: Updated in the recent commits.

File: files/tk_metrics (hunk 2 above)
Comment (jarretlavallee) [reply files/tk_metrics]: Updated in the recent commits

Comment (Sharpie) [reply files/tk_metrics]: I feel like the volume of data is generally large enough that post processing tools are needed to summarize it. `grep` is very convenient, but you can't run `grep` on a data archive that never got uploaded because it was too large for a user to deal with effectively.

If the space savings was smaller, say 5%, then it cold be worth keeping the pretty printing. With a savings of nearly 50%, the opportunity to cut the size of the data in half seems like something we need to take given the difficulty folks have reported with wrangling large support scripts and metrics archives.

Comment (jarretlavallee) [reply files/tk_metrics]: Absolutely agreed. I updated the documentation to provide a way to enable the `grep` workflow and made the change to remove the pretty printing in the latest commits.

**Comment by jarretlavallee** [comment]
Rebased onto master

File: README.md [hunk 3]
```diff
@@ -108,6 +108,13 @@ Optional String: Allows you to override the command that is run to gather metric
 
 ### Grepping Metrics
 
+The metrics come in a JSON hash on one line. In order to convert the metric files into a pretty format, they can be processed with `python -m json.tool` like below.
+
+```bash
+cd /opt/puppetlabs/puppet-metrics-collector
+for i in <service_name>/127.0.0.1/*.json; do echo "$(python -m json.tool < $i)" > $i; done
```
Comment (RandellP) [inline README.md]: Reid made a change so that the real hostname will show up instead of 127.0.0.1   so this example may need to change to represent that.

File: README.md [hunk 4]
```diff
@@ -108,6 +108,13 @@ Optional String: Allows you to override the command that is run to gather metric
 
 ### Grepping Metrics
 
+The metrics come in a JSON hash on one line. In order to convert the metric files into a pretty format, they can be processed with `python -m json.tool` like below.
```
Comment (RandellP) [inline README.md]: Why the change to one-liners?  It was nice to be able to read the files without extra processing.

Comment (jarretlavallee) [reply README.md]: The oneliner changes are to accommodate a change to not pretty-printing the JSON files. We are trying to minimize the amount of space that these metrics take on disk, but more importantly when being sent up to support. We can reduce our collection by hundreds of megabytes using this method as well as uncompressed processing size.

@Sharpie did an analysis of the space savings. 

> I did some benchmarking and it turns out that dropping the extra whitespace created by pretty_generate is going to save us significant amounts of space.
>
> For a monolithic master with 8 compilers and one deployed Code Manager environment, the daily archive size was 16 MB. Removing the file-sync-storage-service and pe- datasets drops that to 13 MB.
>
> Removing the whitespace created by pretty_generate and keeping the data drops the archive size to 7.8 MB.
>
> Removing both the whitespace and the data drops it to 5.6 MB. Disk space required for uncompressed files was reduced from 200 MB to 50 MB once everything was removed.
>
> I would expect the savings from removing file-sync-storage-service to increase rapidly as the number of environments grows, but killing off the whitespace is still a significant reduction in size.

File: README.md (hunk 3 above)
Comment (jarretlavallee) [reply README.md]: Good to know. Updated in the lasted commit.

File: README.md (hunk 4 above)
Comment (RandellP) [reply README.md]: ok, thanks for the data. :)   Wish there was an easy way to convert on the command line for ruby... since we know that is already installed.

Comment (jarretlavallee) [reply README.md]: Agreed. Would it be better off if I included something in `bin/puppet-metrics-collector` to do this conversion instead of the one-liner? Then the user could run something like `./bin/puppet-metrics-collector pretty-print`.

Comment (RandellP) [reply README.md]: That is a great idea.  But while I would like it.  The real question is do the customers actually care.  Cause if they don't ever look at the raw files then it isn't worth adding the maintenance burden.   Of course on top of that, getting the info on if  the customers care might be more work than just doing it. :)

**Comment by jarretlavallee** [comment]
Two approvals, so I am merging this.
